File: module_1/python/src/producer_app.py
import logging.config
import random
import time

from confluent_kafka import Producer

from models.message import Message
from settings.logging import LOGGING_CONFIG

logging.config.dictConfig(LOGGING_CONFIG)
logger = logging.getLogger("producer_app")


def delivery_report(err, msg):
    if err:
        logger.error(f"Error while producing: {err}")
    else:
        logger.info(
            f"Message sent successfully: topic='{msg.topic()}', partition={msg.partition()}, offset={msg.offset()}"
        )


def produce():
    """
    This producer sends messages to Kafka with "at-least-once" delivery guarantee.
    """
    producer_conf = {
        "bootstrap.servers": "localhost:29092,localhost:39092,localhost:49092",
        # Настройка подтверждения доставки:
        "acks": "all",  # Гарантия «at-least-once»
        "retries": 3,   # Поведение при сбоях
    }
    producer = Producer(producer_conf)

    try:
        for i in range(1, 11):
            logger.info(f"Preparing to send message number {i}")
            message = Message(f"user123{random.randint(1, 100)}", "Hello Kafka!")

            try:
                serialized_message = message.serialize()
                producer.produce(
                    topic="test-topic",
                    key=message.user_id.encode("utf-8"),
                    value=serialized_message.encode("utf-8"),
                    callback=delivery_report,
                )
                producer.poll(0)
                time.sleep(0.6)
            except Exception as e:
                logger.error(f"Error while producing: {e}")
    finally:
        producer.flush()


if __name__ == "__main__":
    produce()

File: module_1/project/java/src/main/java/org/example/ProducerApp.java
package org.example;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Properties;
import java.util.Random;

/**
 * This producer sends messages to Kafka with "at-least-once" delivery guarantee
 */
public class ProducerApp {
    private static final Logger logger = LoggerFactory.getLogger(ProducerApp.class);

    public static void main(String[] args) {
        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:29092,localhost:39092,localhost:49092");
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());

        // Настройка подтверждения доставки
        props.put(ProducerConfig.ACKS_CONFIG, "all"); // гарантия «at-least-once»
        props.put(ProducerConfig.RETRIES_CONFIG, 3);  // поведение при сбое

        for (int i = 1; i < 100; i++) {
            logger.info(String.format("Preparing to send message number %s", i));
            try (KafkaProducer<String, String> producer = new KafkaProducer<>(props)) {

                Message message = new Message("user123" + new Random().nextInt(1, 100), "Hello Kafka!");
                ProducerRecord<String, String> record = new ProducerRecord<>("test-topic", message.getUserId(),
                        Message.serialize(message));

                producer.send(record, (metadata, exception) -> {
                    if (exception == null) {
                        logger.info(String.format("Message sent successfully: %s", metadata.toString()));
                    } else {
                        logger.error("Error while producing: ", exception);
                    }
                });
                Thread.sleep(600);
            } catch (Exception e) {
                logger.error("Error while producing: ", e);
            }
        }
    }
}
